### PR TITLE
CLI Client-side Autocomplete

### PR DESCRIPTION
**Pull request by Cleric-K**
Executes silently `help`, `dump`, `get` on CLI open. Parses the output
and builds autocomplete cache. Autcomplete hints are displayed in
popup lists.

Currently autocompletion is supported for:
* primary commands
* get/set setting names and lookup values
* resource names

image

### CODE REVIEW

**Comment by McGiverGim** [comment]
Interesting... How much time takes the execution of this commands to prepare the autocompletion? It does not conflict with the old autocompletion?

Great job!

File: src/js/tabs/cli.js [hunk 1]
```diff
@@ -404,4 +433,305 @@ TABS.cli.cleanup = function (callback) {
...
+
+/**
+ * Encapsulates the AutoComplete logic
+ *
+ * Uses: https://github.com/yuku/jquery-textcomplete
+ * Check out the docs at https://github.com/yuku/jquery-textcomplete/tree/v1/doc
+ *
+ * @param {jquery} $textarea
+ */
+function AutoComplete($textarea) {
```
Comment (McGiverGim) [inline src/js/tabs/cli.js]: If possible, I think it's better to move the AutoComplete to it's own JS file.

**Comment by Cleric-K** [comment]
On F3 about 3sec
On F4 less than a second.

Currently this autocomplete overrides the native autocompletion. If for some reason the parsing fails it fallsback to the native.
I was thinking it could be included as option in the upper right gear menu, but the generation is quick enough that I don't think anyone will need to turn it off.

The get/dump/etc. commands are slow in the CLI because of the DOM itself. Otherwise the serial stream is much quicker. Since I suppress output while generating the cache, the process is as fast as the serial link.

File: src/js/tabs/cli.js [hunk 2]
```diff
@@ -404,4 +433,305 @@ TABS.cli.cleanup = function (callback) {
...
+            builder.state = stateObject.state;
+            TABS.cli.sendLine(stateObject.command);
+            TABS.cli.sendLine(sentinel);
+        }
+        else {
+            // done
+            cache.settings.sort();
+            cache.commands = Object.keys(cache.commands).sort();
+            cache.resources = Object.keys(cache.resourcesCount).sort();
+            writeToOutput('Done!<br># Press Tab for command suggestions<br># ');
```
Comment (McGiverGim) [inline src/js/tabs/cli.js]: The texts of the CLI that come from the firmware are in English, so I think maybe is better to let this in English too but I'm not sure. @mikeller what do you think? Is a good idea to translate this?

**Comment by mikeller** [comment]
Looks interesting, good work @Cleric-K!

In general, I am not a fan of injecting stuff into console apps - this introduces uneeded dependency, and will (for example) break if configurator is connected to a custom build that acts different to what configurator expets. For this reason I think making it optional and allowing the user to disable it should be done.

Also, adding libraries by adding the library to the repository is deprecated, and should be avoided if possible. Adding an npm dependency on `jquery.textcomplete` (or rather `textcomplete`, since the former is deprecated), and then referencing the minified version included in that package, is the preferred way.

**Comment by Cleric-K** [comment]
OK.
So I'll look into the npm stuff. I'm still little confused about the bridging between node.js and browser/js.

I chose `jquery.texcomplete` (even though is deprecated) because it has a little clearer api (especially since we use jquery). The new `textcomplete` uses classes but I'll take a look.

I'll add option for disabling in the preferences.

@McGiverGim About moving to a separate file: `AutoComplete` calls methods in `TABS.cli` and also globals in `cli.js` (for example `writeToOutput()`). Would it be OK if these functions are called across files?

**Comment by mikeller** [comment]
@Cleric-K:

Sticking with `jquery.textcomplete` is ok if this is a better fit for us - I just wanted to point out that it is listed as deprecated.

Re node / npm, the thing to keep in mind is that we still have a considerable percentage of users using the Chrome Web App, and we'll need to stay compatible with it at least for now. This means we can't use `require`. Luckily npm packages often include a self-contained, often minified, 'browser ready' version that can be directly referenced from web pages, and these 'browser ready' versions can be included in the Chrome Web App (and built into the NW.js version) just like they are included in JS called in web pages.

**Comment by McGiverGim** [comment]
About the calls to methods of CLI, if they are not too many of them, maybe is better to pass them as callbacks or something similar, to isolate the AutoComplete class. But at first look, I don't see if it will be easy or only complicate things. 
The idea is not too let grow too much the JS file, it will grow continuously making it hard to maintain.

**Comment by Cleric-K** [comment]
Got it.
I guess it should go into the `workers` directory?

**Comment by McGiverGim** [comment]
Maybe, but take a look first to see if the final code will be better. I'm on mobile so is difficult to me to see the implications. My message was only a suggestion, if it helps to isolate the code, perfect, but if it will only complicate things, maybe is better to maintain it as is.

**Comment by mikeller** [comment]
Agreed with @McGiverGim re using callbacks. But in general, splitting up the code in files by function is already an improvement, as it forces developers to think about APIs and stems bad practices such as overly large closures.

**Comment by Cleric-K** [comment]
OK. I'll look at it tomorrow.

In fact I started by adding the code directly to `TABS.cli` but later decided to separate it as a class just to package the autocomplete data into a more monolithic whole (instead of adding bunch of properties to `TABS.cli`) - now everything is inside the `TABS.cli.autoComplete` object.
But other than that they are still quite interrelated - autocomplete calls the `send` and `write` functions, the `read` function of `cli` has to decide whether to send incoming data to `autoComplete` for cache building or direct to output... so it's hard to imagine one without the other.
If it was written in Java I guess `AutoComplete` would be a nested class.

**Comment by Cleric-K** [comment]
* `CliAutoComplete` code moved in a separate file
* Preference for enabling (on by default)
* `textcomplete` through npm. I'm still using `jquery-textcomplete`. I tried `textcomplete` and it works but there's one very annoying issue. The jq version has the nice feature that the popup list can be scrolled not only with `UP`/`DOWN` keys but also with `PAGEUP`/`PAGEDOWN`. This is missing in the newer version and it's very annoying since we can easily get long lists (with `get`/`set`) and trying to move with `PAGEUP`/`PAGEDOWN` moves the cursor in the `<textarea>` instead. We can work around this by conditionally `preventDefault()` for these keys if the list is open but it's getting ugly. Plus the jquery version looks a little nicer in code. Thus I suggest we stick with the jq version at this point. The Api is practically the same so if something forces us in the future to upgrade only few lines will have to be changed.
* Tests fixed

So, do we need to localize the `Building ...` message in the CLI window?
The FC's cli output is always in English so we may conform with that.

File: locales/en/messages.json [hunk 3]
```diff
@@ -4154,5 +4154,8 @@
     },
     "flashTab": {
         "message": "Update Firmware"
+    },
+    "cliAutoComplete": {
+        "message": "Client-side CLI AutoComplete"
```
Comment (McGiverGim) [inline locales/en/messages.json]: This is for "normal" users. They don't understand "client-side" or similar. I think is better to put something like "Advanced AutoComplete in CLI" or "Contextual AutoComplete in CLI", that will be better understood.

Comment (Cleric-K) [reply locales/en/messages.json]: I agree. I was going to ask exactly about this but forgot.
@mikeller what would you propose?

I'll push another commit in a while. Found few bugs that affect if turning on/off this feature while the CLI is opened.

Comment (mikeller) [reply locales/en/messages.json]: I'd also think 'advanced auto-complete' is the right thing from the users' perspective - they don't care much where it is generated.

File: package-lock.json [hunk 4]
```diff
@@ -91,7 +91,7 @@
         "string-width": {
           "version": "2.1.1",
           "resolved": "https://registry.npmjs.org/string-width/-/string-width-2.1.1.tgz",
-          "integrity": "sha512-nOqH59deCq9SRHlxq1Aw85Jnt4w6KvLKqWVik6oA9ZklXLNIOlqg4F2yrT1MVaTjAqvVwdfeZ7w7aCvJD7ugkw==",
+          "integrity": "sha1-q5Pyeo3BPSjKyBXEYhQ6bZASrp4=",
```
Comment (mikeller) [inline package-lock.json]: Erg, this is an artefact of an old and broken version of npm being used, or at least cached files downloaded by such a version. Update npm to latest, then follow https://github.com/npm/npm/issues/17749#issuecomment-322627101 to fix it, revert the changes to `package-lock.json`, and do an `npm i` again.

**Comment by Cleric-K** [comment]
* Preferences text changed to *Advanced CLI AutoComplete*
* `package-lock.json` fixed
* Added `beeper` and `feature` command support. I doubt many people configure these through CLI, but anyways, it was just a few lines of code to support them.

**Comment by Cleric-K** [comment]
After hundreds and hundreds tests, yesterday I got `Failed` building once. I couldn't understand why it happened, neither I could reproduce it.
The strange things was that I got in the console the `Failed!` message, but there was no `Building...` message (which should precede it). I couldn't figure a code path that could lead to such a result.

Anyway, I've added a single _retry_ just in case. If the building fails, it tries silently once more and if that fails too, it produces the `Failed` message and fallsback to native autocomplete.

**Comment by bizmar** [comment]
I tried it today, and while I really wanted this for the last few months while testing dev builds... This implementation does not work for me. Most of the time I want to check a few settings at a time, a lot of features are similary named so a `get` brings them all up at once, like `get RPM`, `get dyn`, `get d_min`.. this implementation  forces you to select one full name.. I would like auto complete to be invoked by pressing the tab key, or similar.

**Comment by Cleric-K** [comment]
@bizmar I'm glad you're bringing this up.
So far the discussion was focused more on the source code level. I was expecting to comment on the actual UI experience, because I'm far from the thought that it is perfect.

I don't know if @McGiverGim and @mikeller had the chance to actually test it already.

Different parts of commands have different completing behavior.
* primary command - For this I have chosen to autocomplete only by _prefix_, since most commands are single worded and it would be rare someone for example to need execute `status` but start typing `tus`. Also, currently the list won't open by itself - only on Tab (I guess that's what you expect from `get`)
* `get`/`set` settings names - for these I chose to show suggestions by matching _anywhere_ in the string, since that's how people are used to work already (when getting values like you say - `get rpm`, etc.).
For this part I chose the autocomplete to popup automatically after the third character. My sole motivation for this is just to make the user aware that such thing exists. Many people have no experience  with consoles (Linux for example) and they simply would not know that such behavior can be expected from the Tab key.
* `set` values - this pops up automatically. Especially for the string lookup values I think there's no downside to that. Matching is also _anywhere_.
* `resource`, `feature`, `beeper` - these show suggestions after the first character. Again I don't see any downside to this.

So I perfectly understand your experience with the `get` command. Technically it does not _force_ you to choose single command - you can hit `ESC` and it will close the popup, then hitting `ENTER` will execute the `get` with incomplete name. Of course this is cumbersome and requires intuition about the `ESC` key which most users in our GUI age probably don't have.

If you want to try the behavior in the way you need it, open `src\js\CliAutoComplete.js`, go to lines `239` and `250`, and change the last argument of the two`searcher()` calls (currently `3`) to some large number. This is the min chars that have to be input before the popup is opened.

As a middle solution we can:
* make `get` _not_ open automatically
* make `set` open automatically after third character. This makes more sense, since `set` requires single name (although it can be executed with partial name and it behaves like `get` but probably not many use it in this way).

I'll be happy if everyone gives his opinion.

**Comment by McGiverGim** [comment]
I've been out today. Tomorrow I will try to test it and give my opinion.

**Comment by mikeller** [comment]
@Cleric-K: No, I have not yet had the chance to test this - kind of busy with upcoming Betaflight 4.0 at the moment.

On a related note, I think that this is a great improvement, but I don't think that it is a good idea to try and squeeze this into the upcoming 10.5.0 release that we'll have to do to go out to support Betaflight 4.0 when it is released. There still are some details that need to be worked out, and merging this early after 10.5.0 is out will give us a lot of time to get and incorporate user feedback.

**Comment by McGiverGim** [comment]
@mikeller maybe we can include it in the next release of the Configurator (for Betaflight 4.0), but if the user experience is not polished 100%, maybe we can disable it by default (now is enabled). In this way we can let the users test it and get feedback and suggestions. And enable it for Betaflight 4.1 if at that moment we are happy with it.

I will test it in some hours and give my own feedback :)

**Comment by Cleric-K** [comment]
Maybe I'll just push one last change as @bizmar suggests: to have `get` show popup only on `Tab`. While `set` may remain to auto popup after 3rd char.
What do you think?

**Comment by bizmar** [comment]
That would be perfect, maybe suppress the indexing message on F4 as it's fast enough that it's not noticeable, and only show if an error occurs

**Comment by McGiverGim** [comment]
I've been testing, good work!!! I like it :)

The  `get` and `set` must behave equal. I don't think we must throw the auto suggestion only with tab in one and with tab and automatic for the other.

The different solutions to fix the get problem, can be:
1. Don't throw the auto suggestion in an automatic way, only with pressing `tab`. The message in the terminal is clear.
2. Throw the auto suggestion but only after some time (1 or 2 seconds) without the user entering nothing. In this way, if the user wants can throw the partial `get` command without any problem.
3. Add as first suggestion of the `get` command the `get` partial entered until this moment by the user.. 
4. Don't select the first suggestion automatically. I don't know if this is possible. In this way if the user does not press some cursor key to select one suggestion, he can throw the current partial `get` without problems.

I don't know what the library admits. I will implement 4 if possible in all the cases. 3 is a secondary option if 4 is not possible.
After that, I prefer 1, but 2 can work too, specially with 4 or 3.

My ending thoughts:
- Implement 4 if possible.
- Only throw the auto completion with tab (1) or after some time without entering nothing (2).

This is only my opinion, I don't know what others think...

**Comment by Cleric-K** [comment]
OK, here're my thoughts:

`4` would be somewhat challenging. There's no such option out of the box. I looked at `textcomplete` source. It keeps private variable with the selected index which would be impossible to reach from outside without modifying the library.
Option `3` is quite clever and does not modify the cli habits of the user. I vote for it.

About the timeout - this can be useful to remind the user that autocomplete exists but other than that it might not be as practical. I doubt anyone will use it normally - that is wait for the list to appear instead of hitting `Tab`. Plus, it will require some more attention on the different event types - activating and deactivating the timeout at the proper times, in order to avoid suggestions popping up out of the blue.

I've pushed a version with option `3` implemented.
Also, I've moved the `Press Tab to ...` hint to the `placeholder` attribute of the `<textarea>`. I think it's much more visible there plus it's localizable.

@bizmar I don't know if mcu type is readily available in the configurator (have to be somehow inferred from Target). Anyway, now with the hint out of the way, I think the `Building ...` message is not that intrusive even on F4.

**Comment by McGiverGim** [comment]
I've tested your version with 3 included, to me is good, but I don't know if it was intended or not, it seems it has 1 included too (until I press tab it does not work).

**Comment by Cleric-K** [comment]
Oh, I see. I've misunderstood, so you say that in `3` we still get auto popup after 3rd char? Got it.

I also noticed something about `resource` - I'll add the `show` and `show all` arguments in the suggestions.
Also `list` for `beeper` and `feature`.

**Comment by McGiverGim** [comment]
I think the option 3 is good with automatic autocompletion. Without it is not needed because is the user who asks for the suggestions pressing tab and can close it with esc. 

I prefer the user asking for them, but if you and others think that is better to pop out it automatically, then the option 3 is needed and I will not oppose to that neither.

**Comment by Cleric-K** [comment]
OK. Let's try `3` with auto popup after 3rd char.

Now I'm wondering about `resource`. Its api got changed at some point from `resource list` to `resource show [all]`. If I have to support both I'll have to check against the bf version. I can get that from the `dump` command.

**Comment by McGiverGim** [comment]
You can check the MSP version API. It will be enough and is read when connecting. Check any PR to see how it is done.

**Comment by mikeller** [comment]
@Cleric-K: Good point - you should probably also check the flight controller firmware type, and not even try to extract data if it isn't Betaflight - users are using Betaflight configurator to communicate with other things that aren't Betaflight by using the CLI.

**Comment by Cleric-K** [comment]
But it has to be communicating over MSP, right? That is - the comm always starts with MSP, you can't use the configurator as general serial terminal?
My point is, can I assume that the MSP version, etc. will be populated in all cases by the time cli is opened?

**Comment by McGiverGim** [comment]
In the Configurator, yes. When the user connects the MSP version is read and printed in the upper log. I don't see a way to go to the CLI tab without pressing the Connect.

**Comment by mikeller** [comment]
@Cleric-K:

> But it has to be communicating over MSP, right?

There are a lot more projects that use the Multiwii Serial Protocol when communicating, and some of them only implement a very basic subset of it - but pretty much all of them support querying the firmware type and API version, and dropping into CLI. Whenever communicator is connecting to something that isn't Betaflight, it will drop into a compatibility mode that does not have any of the tabs except for CLI.

**Comment by McGiverGim** [comment]
I thought that you only will enable the auto completion for Betaflight. You can check in the values stored from MSP if Betaflight and version. If something goes wrong deactivate the feature.

**Comment by bizmar** [comment]
If you click on "save to file", it will save the hidden dump, along with the things (probably a diff) that you meant to save, i'm using a version from a few days back, so this may have been fixed.

**Comment by Cleric-K** [comment]
Nice catch, @bizmar !
Will add the fix to the other stuff I'm working on. Tomorrow I hope I'll be able to push the rework.

**Comment by Cleric-K** [comment]
* AutoComplete is enabled only if fc identifier is `BTFL`
* `resource` command overhauled. All arguments are hinted. `list`/`show` subcommands hinted based on firmware version
* more robust building process. Sentinels are generated per building, otherwise if we retry the building (because we are not getting response), but then we receive the abnormally delayed data from the first build we would not be able to tell it apart.
* building dialog is not saved when exporting the cli output to file. Furthermore a small fix was applied that affected the native autocomplete - if autocomplete was performed which included corrections with backspace, the BS characters (0x08) sent from the FC get recorded into the saved file.
* support for `mixer` command
* added `list` subcomand to `feature`/`beeper`
* added the `ALL` argument of the `beeper` command to the hint list

**Comment by McGiverGim** [comment]
Good job! A lot of fixes... I will test it later :)

**Comment by McGiverGim** [comment]
Tested, and to me is good :)

Let's see what users think when merged, but I'm pretty sure they will like it, it is a great feature...

**Comment by Cleric-K** [comment]
Great, glad it works for you too. So far I've been using it and seems OK.

The main thing is that, with everyone's ideas here, from behavior perspective, the way it works now is completely non-intrusive - that is, if the user ignores the visual element, and blindly uses the keyboard as before, the end result should be exactly the same.

So it is up to you to decide if it is to be merged in `10.5.0`.

File: src/js/CliAutoComplete.js [hunk 5]
```diff
@@ -0,0 +1,482 @@
+'use strict';
+
+/**
+ * Encapsulates the AutoComplete logic
+ *
+ * Uses: https://github.com/yuku/jquery-textcomplete
+ * Check out the docs at https://github.com/yuku/jquery-textcomplete/tree/v1/doc
+ */
+const CliAutoComplete = {
```
Comment (mikeller) [inline src/js/CliAutoComplete.js]: I am normally trying to keep the configurator coding style as close as possible to C coding style. This seems to make sense since the 'main' project is the firmware which is in C, and we want to make it as easy as possible for contributors to the firmware to make the changes to configurator that are required to match their changes in the firmware themselves.

This approach would call for constants to be named in MACRO_CASE.

File: src/js/CliAutoComplete.js [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+        // got sentinel
+        var command = builder.commandSequence.shift();
+
+        if (command) {
+            // next state
+            builder.state = 'parse-' + command;
+            this.sendLine(command);
+            this.sendLine(builder.sentinel);
+        }
+        else {
```
Comment (mikeller) [inline src/js/CliAutoComplete.js]: Can we put these as `} else {` please?

File: src/js/CliAutoComplete.js (hunk 5 above)
Comment (Cleric-K) [reply src/js/CliAutoComplete.js]: Sure,
Actually I've never used the `const` keyword in `javascript` before. I just saw that it is in use in the existing code and decided to go along.
About the MACRO_CASE - I'm also used to this, but I thought that in `javascript` `const` is used more in the sense of `final` as in `Java`.

What would you recommend? Simple to do it as `var`s, unless it's a real constant primitive (which will be uppercased)?

File: src/js/CliAutoComplete.js (hunk 6 above)
Comment (Cleric-K) [reply src/js/CliAutoComplete.js]: Will do.

**Comment by Cleric-K** [comment]
> However, I think it is a better approach to put this into 10.5

I guess you meant `10.6`?

Either way I agree - let's do this slowly and properly.

I think it will be useful to discuss the scope of the autocomplete feature (in the context of the `timer`, `dma` commands).

So far I've focused on the completion of keywords and constant lookup values. When I was adding support for the `mixer` command, I also did experimental `mmix` support. The result, however, could not be called exactly _autcomplete_, since the popups do not offer any completion but serve more as _tooltips_ (similarly to the `Allowed range: ...` hint for numeric settings). So `mmix` was working by displaying hints in a similar fashion: when the cursor was in the place of the first argument a `<motor>` popup would appear. When the cursor moves to the second argument `<roll>` popup appeared and so on. But, of course, pressing tab/enter would not complete anything - it's just a tooltip. So in the end I have not included it in the commit.

I had similar dilemma when writing the `resrouce` support. I could easily make cache of the pin names and make autocomplete for them, but in the end decided it was not that useful. Pin names are three characters and you have to write them all anyway or use multiple arrow keystrokes, to distinguish between the other autcomplete options. The only benefit would be that you're able to see what are the available pins, but that's probably also not that useful, since if you are writing the command you already know what pin you would like to write. That's why I resorted to the simple `<pin>` tooltip.

So the question about all the other commands like `dma`, `timer`, etc. is what kind of autocomplete we would like to implement? So far I think it can be only in the form of tooltips that hint the argument type at the cursor position.

I can implement `dma` and `timer` support but still I'm not very clear about how they are used. I have experience with STM32 programming, but I haven't looked into bf source to see how these commands translate to the low level. For example, I'm not clear on how `dma` is attached to a _pin_. On the low level a dma stream is attached to a request _channel_ - which is triggered by certain peripheral. Does this mean that the `dma` command infers the peripheral from the pin number as it's described by `resource`?

**Comment by mikeller** [comment]
@Cleric-K:

Yes, correct, I meant 10.6.

Point taken about not adding a list of available pins as an autocomplete. My idea was to show all the pins that can be used for any given target (which is defined in `TARGET_IO_PORT<x>` in the target configuration. But you are right - if you don't already know what pin you want to assign a function to, you are not in a position to enter a `resource` command.

The same obviously goes for the `timer` and `dma pin` commands. The only one that does make sense to offer autocomplete for is the `dma` first argument to the `dma` command, with either `pin` or the name of a dma consuming resource as the input - I for my part always struggle to remember how the resources are called.

Re `dma` attachement to pins: You are obviously right, DMA resources that are used to drive pin output are not assigned to the pin at hardware level, but rather to the timers that are used to control the DMA timing, which in turn are assigned to pins. But, in order to keep the configuration compact, and because there is no way to use a DMA stream that is assigned to a timer that is not assigned to a pin, the dma to timer assignments are stored in the same list that the timer to pin assignments are stored.

**Comment by Cleric-K** [comment]
* fixed `} else {`s
* replaced `const` with `var`

At this point I think we have to have the `timer/dma list` subcommands implemented in order to get the device names. If I read the source right the resource _owner_ names (which `dma` uses) are not exactly the same thing as the `resource` names that we already have implemented.
In other words, leaving the pins aside for the moment: is there anything in the _current_ implementation of the `timer/dma` commands which can be parsed and used to make somewhat useful autcomplete for them?

About the pins: I think one aspect where pin suggestion could be useful would be if we can somehow filter the _allowed_ pins for a resource. At this point it seems nothing enforces what pins can be selected for what hardware. For example, at the hardware level only certain pins can be used for say, `UART3 TX`. I don't know if these constraints are present as definitions in bf.
Thus it can be helpful if for example one starts with `resource MOTOR 2 ` and then only pins that can be used for that resource show as suggestions.

Clearly, this does not seem to be possible at this point. But it might be an idea for future work. There can be a hidden "plumbing" cli command, designed to be parsed by autcomplete, which provides such information.

**Comment by mikeller** [comment]
Good point about the `timer/dma list` commands - they are on my to-do list for after 4.0 is out.

Re allowed pins, the firmware does indeed not have a concept of this, the only restriction is the list of overall allowed pins from `TARGET_IO_PORT<x>` (which are also not really applicable for Unified Targets, since all pins available on the MCU are allowed for these. I also do not think that this will change any time soon - there simply isn't a good enough use case for it, as what pins can effectively be used for any given function will always depend on external components and availability on a pin / port for the given board, so the concept of a user 'freely' choosing a pin is just not there.

**Comment by Cleric-K** [comment]
Got it.

Well, then I guess this is pretty much what we can do at this stage.
We can add `timer/dma` support when the `list` commands are implemented.

Anyone has any further ideas that we can implement?

**Comment by McGiverGim** [comment]
Are there any changes pending? I think is a good moment to include it in master, with a ton of time until the new release to gather feedback from users :)

**Comment by McGiverGim** [comment]
@Cleric-K I think I have found a bug, this are the steps:
- Type `set debug`
- Move to `debug_mode` and press TAB.
- Type `gyro` 
- Move to `GYRO_SCALED`
- Press enter

Nothing appears in the editor. The command has been sent, but I can see it.

**Comment by Cleric-K** [comment]
I've pushed rebase on current master. Otherwise for the moment I don't have anything else to add.

@McGiverGim This is intended behavior. I've tried to make `Enter` execute command directly _if_ the argument being completed is the _last_ argument of the command. I thought it makes sense because otherwise you have to press enter twice. You can still complete without sending, if you want, with `Tab`.

**Comment by McGiverGim** [comment]
The problem is that it does NOT appear in the editor. The command is sent but you can't see it. At least in my computer.

**Comment by Cleric-K** [comment]
That's weird 🤔
You don't see only the echo of the command? Or also the output `debug_mode set to ...`

It works fine here:
image

What fc/bf ver are you testing with?

**Comment by McGiverGim** [comment]
Confirmed:
- MATEKF722SE -> Has the problem
- OMNIBUSF4SD -> Works perfectly

**Comment by Cleric-K** [comment]
Very interesting. I don't have F7 board to test with.

How should we proceed?
Do you have the time to do some tests?

First some questions:
1. You said that, but just to confirm: the command _is_ executed, right? Only not displayed?
2. Does this happen _only_ when executing with autocomplete on `Enter`? For example if you autocomplete the `set debug...` with `Tab` and press `Enter`, do you get the output?

Is it possible to edit `tabs/cli.js` and add `console.log(data[i] + ' ' + currentChar);` on line `322`?
This should give us an idea if we get echo from the fc at all.

If we do get the the characters, the only thing I'm seeing at this moment is the: `writeLineToOutput(text)`  func in `cli.js`. This seems to be the only capable of suppressing the output. We should place breakpoint on `isBuilding()` in `CliAutoComplete.js` and see what `this.builder.state` is.

**Comment by McGiverGim** [comment]
Some things:
- Yes, the command is sent, only not displayed
- It does not happens always, sometimes it works, but I can reproduce it without problem.
- If I select another option in the autocomplete, it seems to work, by the moment only gyro_scaled produced the problem. I have tried others but they worked. I'm pretty sure this will not be the only one, but I can find others at this moment.
- If I autocomplete with TAB before the enter, it fails in the same manner.
- If I write the command by hand, without autocomplete, it works, and it shows this command and all the previous that failed.

Perhaps is some kind of buffer, that is not wrote to the output?

**Comment by McGiverGim** [comment]
Added the line in the 322, it does no execute it. It only executes it if I write something by hand to the output.
Maybe the output buffer is not being sent until I write something that works?

**Comment by McGiverGim** [comment]
I can confirm, when I execute something by hand, all the commands introduced arrive to the line 322 (method read). So it seems something is being stored in a buffer, but not sent for some reason.

**Comment by Cleric-K** [comment]
The question is if the output is held at the mcu or is held at the configurator serial code.
There's something which triggers the output to be flushed in one case and the other (manual or autocomplete).

Can we try before line `255` (before triggering the event) in `CliAutoComplete.js` to add:
`$textarea.val( $textarea.val() + "\r" )` ?
To see if CR helps trigger the flushing

**Comment by McGiverGim** [comment]
No, it does not work. Tested \r and \r\n.

The enter event is triggered, and executed the `textarea.keypress(function (event)` and the promise that includes. So it seems it is being stopped at some part of the serial. Maybe it has a buffer and until full (or until some special key) it does not send it?

**Comment by Cleric-K** [comment]
Weird stuff ...

I also thought about special byte that trigger serial flushing, that's why I suggested adding the `\r`... but now I'm really at loss, because I don't see what could be the difference between triggered after autocomplete and normal input.

The last thing I can think of is that autocomplete adds `space` after completing the last word. Can you try for example autocomplete with `Tab` then erase last space and `Enter`?

**Comment by McGiverGim** [comment]
No, it does not work neither removing the space. And is strange because other commands work.
- `set debug_mode = GYRO_SCALED` -> command is stopped
- `set debug_mode = GYRO_RAW` -> command is sent

**Comment by Cleric-K** [comment]
But I don't understand what's the difference when you send `GYRO_SCALED` by entering manually and after complete. Very weird. From the point of view of the `sendLine()` func there should be no distinction.

So if you enter manually char by char the exact same command (including case and space at end) it works, but if you autocomplete with tab and enter it won't?

**Comment by McGiverGim** [comment]
I don't know if I'm looking at the correct place, but in `serial.js`, line 308, the content of `data`, with autocomplete disabled (and command working is):

image

But with your autocomplete, it has:
image

As you see, they are very different. If I enter the command char by char OR I use the autocomplete but I press ESC before enter (to hide the autocomplete suggestion), then it sends again the first image, and it works!!!

**Comment by McGiverGim** [comment]
The size of the array, 29 or 30 is for the latest space (code 32) but this Int16Array between the two Int8Array is what is strange to me...

**Comment by McGiverGim** [comment]
I have found where it happens, but I don't know why:

`cli.js` line 412, the content of `line` and the `bufferOut` after initialization is:
image

And after the `for` loop:
image

Is like the charset was not correct or something similar...

**Comment by Cleric-K** [comment]
Maybe the Int16 view appears only if the byte count is even?

**Comment by Cleric-K** [comment]
Yes. I tested `new ArrayBuffer(count)` creates views based on the buf length. If `count` is multiple of `2` it adds `int16` view. If it's multiple of `4` it adds `int32` view also.

So it seems that the interval at the end bugs it (although you said that if you manually remove it still does not work?)

Here's something to try: at the place where you added the `"\r"` line `255`
`$textarea.val( $textarea.val().trim() )`
This should remove any trailing spaces and the argument to `sendTo()` should be exactly the same with or without autocomplete.

**Comment by McGiverGim** [comment]
Precisely I was testing the same than you proposed... and it works!!!!

It seems that when I removed it manually, pressing enter (whitout pressing ESC) executed the autocompletion again and added the space again, so this was the reason to not work. And it seems to produce problems only when the size of the command is 30 (I don't know why).

I suppose some kind of bug in the serial.js or similar, but with this, it works for sure.

**Comment by McGiverGim** [comment]
I tested adding it to the `cli.js`, line 199, changing:
```
                var out_string = textarea.val();
                self.history.add(out_string.trim());
```
by
```
                var out_string = textarea.val().trim();
                self.history.add(out_string);
```
and it works too. But it seems prepared to work with multiline input (it does an split by `\n` later), so maybe is better to fix simply in your part.

**Comment by Cleric-K** [comment]
There, I've added the fix in my part.
It's still curious why this happens... So if now you manually enter the command with one `space` at end, it won't work, right?

Here are some debug modes:
```CYCLETIME
ANGLERATE
SCHEDULER
DUAL_GYRO
ESC_SENSOR
SMARTAUDIO
RPM_FILTER
GYRO_SCALED
RANGEFINDER
ITERM_RELAX
RX_FRSKY_SPI
RX_SFHSS_SPI
```

sorted by length. Do `RANGEFINDER` and `ITERM_RELAX` have the same behavior?
What about `RPM_FILTER` but with _two_ trailing spaces?

Interesting that there's no problem with OMNIBUS which hints that js code should be sending data properly. I wonder if there could be a bug in bf for F7. It could be something buried in the HAL VCP code ...

**Comment by Cleric-K** [comment]
Also interesting if it has the problem with `RX_FRSKY_SPI` which should be 30 chars even without trailing space.

**Comment by McGiverGim** [comment]
I will test all of this maybe tomorrow, to see what happens. Now I'm out of the computer. But adding spaces to fit length 30 does not produce the problem with others elements, I tested it.

**Comment by McGiverGim** [comment]
Tests:
- Manual adding an space at the end: fails, `set debug_mode = gyro_scaled ` tested with autocomplete disabled and it did not work. Adding two spaces work again.
- RANGEFINDER and ITERM_RELAX with space fail
- RPM_FILTER with two spaces, work
- RX_FRSKY_SPI work, with and without space

So it seems to fail when the command is 29 lenght and one space, 30 in total. 

@mikeller this strange behaviour. Maybe is a bug in the firmware part of Betaflight? Can you made this test with some F7 (I tested with a MATEKF722SE). 

What it seems:
- We send the "bad" commands to the FC, but the FC does not respond. It stores them. When a "good" command is sent, then it executes all the stored and responds to all.
- F4 does not seem to have this problem, only F7.

**Comment by Cleric-K** [comment]
Thank you! Interesting.

Another test that could potentially help a little is opening the serial port in other program, for example PuTTY, activate cli with # and execute the commands.
If it fails here too we can be sure that it is not js fault. It can be effect in windows drivers or at the mcu hal layer.

The pc driver can be further tested on Linux. If raw terminal on Linux has the problem too, we can be quite confident that something happens in the mcu.

**Comment by McGiverGim** [comment]
I will try to test it, but I'm out of home. In one or two days I will test with putty.

**Comment by McGiverGim** [comment]
Other user have tested in putty and it works. In Configurator same problem than me. So it does not seem a firmware problem.

**Comment by Cleric-K** [comment]
Hm.. so that leaves it in nw.js/chrome?
Maybe if we try to add `chrome.serial.flush()` after sending this will confirm ..

**Comment by McGiverGim** [comment]
I will test when I can, we can made a trim in the general code of cli.js, after the split, as a workaround if we can't find the origin. But this must go in other PR, because it's clear that is not related with the auto completion, so by the moment for me your code is ok.

**Comment by McGiverGim** [comment]
Another suggestion: using F3 is slow. I try to paste something in the text box, but it does not respond (it is loading the autocomplete). Maybe is better to disable the text box with a "Wait please until AutoComplete has finished loading..." or similar message, until the autocomplete has finished.

**Comment by mikeller** [comment]
@Cleric-K, @McGiverGim: Yes, this issue looks much like other problems we are having on F7 with the USB stack (see betaflight/betaflight#7873). We'll have to look into replacing the library versions with newer (and hopefully better) versions soon-ish.

**Comment by Cleric-K** [comment]
Agree.
Maybe the proper way to do it would be to make events for start and stop building in the `CliAutoComplete` object and make `TABS.cli` listen for them and disable/modify placehodler of the `<textarea>`. This will be slightly more code but code will be more decoupled (rather than `CliAutoComplete` manipulating the `<textarea>` directly. What do you think?

**Comment by Cleric-K** [comment]
Pushed a variant with events. Using jQuery `.on/.off/.trigger`.

**Comment by Cleric-K** [comment]
Can you elaborate:
on a empty argument and pressed `Tab`, what would you like the list to look like?
`list` + all the features twice - once without and once with `-`?

**Comment by mikeller** [comment]
No, I think having the features once (to enable them), plus `list` and `-` should be enough. Then, after the user entered `-`, the features (sans `list` and `-`) should be listed.

**Comment by Cleric-K** [comment]
There, I think that's what you had in mind.

**Comment by McGiverGim** [comment]
I have added the trim to the general code of CLI here: https://github.com/betaflight/betaflight-configurator/pull/1388, if you want you can remove the trim in the AutoComplete.

**Comment by mikeller** [comment]
Looking good now!

**Comment by Cleric-K** [comment]
Done, `trim()` removed.
